src/storage/v3/storage.cpp

```diff
@@ -56,36 +63,78 @@ using OOMExceptionEnabler = utils::MemoryTracker::OutOfMemoryExceptionEnabler;
+    *vertex = VertexAccessor::Create(&it->vertex, tx, indices, constraints, config, schema_validator, schemas, view);
     if (!*vertex) {
       ++it;
       continue;
     }
-    break;
+    return it;
+  }
+  return std::nullopt;
+}
+
+AllVerticesIterable::Iterator::Iterator(AllVerticesIterable *self, Labelspace::iterator labelspace_it)
+    : self_(self), labels_it_{labelspace_it} {
+  if (labels_it_ != self_->labelspace_->end()) {
+    self_->vertex_accessor.emplace(labelspace_it->second.access());
+    vertex_it_ = AdvanceToVisibleVertex(self_->vertex_accessor->begin(), self_->vertex_accessor->end(), &self_->vertex_,
+                                        self_->transaction_, self_->view_, self_->indices_, self_->constraints_,
+                                        self_->config_, *self_->schema_validator_, *self_->schemas_);
+    if (vertex_it_ == std::nullopt) {
+      end_ = true;
+    }
   }
-  return it;
 }
 
-AllVerticesIterable::Iterator::Iterator(AllVerticesIterable *self, VerticesSkipList::Iterator it)
-    : self_(self),
-      it_(AdvanceToVisibleVertex(it, self->vertices_accessor_.end(), &self->vertex_, self->transaction_, self->view_,
-                                 self->indices_, self_->constraints_, self->config_, *self->schema_validator_)) {}
+AllVerticesIterable::Iterator::Iterator(AllVerticesIterable *self, Labelspace::iterator labelspace_it, bool end)
+    : self_(self), labels_it_{labelspace_it}, end_{end} {}
 
 VertexAccessor AllVerticesIterable::Iterator::operator*() const { return *self_->vertex_; }
 
 AllVerticesIterable::Iterator &AllVerticesIterable::Iterator::operator++() {
-  ++it_;
-  it_ = AdvanceToVisibleVertex(it_, self_->vertices_accessor_.end(), &self_->vertex_, self_->transaction_, self_->view_,
-                               self_->indices_, self_->constraints_, self_->config_, *self_->schema_validator_);
+  while (labels_it_ != self_->labelspace_->end() && vertex_it_) {
+    ++*vertex_it_;
+    if (*vertex_it_ != self_->vertex_accessor->end()) {
+      vertex_it_ = AdvanceToVisibleVertex(*vertex_it_, self_->vertex_accessor->end(), &self_->vertex_,
+                                          self_->transaction_, self_->view_, self_->indices_, self_->constraints_,
+                                          self_->config_, *self_->schema_validator_, *self_->schemas_);
+      if (vertex_it_) {
+        return *this;
+      }
+    }
+    ++labels_it_;
+    if (labels_it_ != self_->labelspace_->end()) {
+      self_->vertex_accessor.emplace(labels_it_->second.access());
+      vertex_it_.emplace(self_->vertex_accessor->begin());
+      vertex_it_ = AdvanceToVisibleVertex(*vertex_it_, self_->vertex_accessor->end(), &self_->vertex_,
+                                          self_->transaction_, self_->view_, self_->indices_, self_->constraints_,
+                                          self_->config_, *self_->schema_validator_, *self_->schemas_);
+      return *this;
+    }
+  }
+  end_ = true;
```

Comment:
Just want to point out this for the reviewers, I was not sure if we should have all accessors opened at once or like this, one at the time. Curerently we open them when we pass to next skiplist, this iterator iterates through map of skiplists, and everytime we jump to new skiplist we need to access it by creating accessor object